torch_geometric/nn/models/lightgcn.py

```diff
@@ -0,0 +1,315 @@
+                graph.
+            edge_label_index (Adj, optional): Edge tensor specifying the start
+                and end nodes for which to compute rankings or proabilities.
+                If :obj:`edge_label_index` is set to :obj:`None`, all edges in
+                :obj:`edge_index` will be used (default: :obj:`None`).
+            weights (Tensor, optional): Pre-computed embeddings or node
+                features which should be used to initialize the embedding
+                layer (default: :obj:`None`).
+        """
+        if edge_label_index is None:
+            edge_label_index = edge_index
+
+        if weights is not None:
+            assert (weights.shape == self.embeddings.shape), \
+                'Pre-computed embeddings weights must match shape of \
+                embedding layer.'
+            self.embeddings.weight.data.copy_(weights)
+
+        out = self._propagate_embeddings(edge_index)
+
+        nodes_fst = torch.index_select(out, 0, edge_label_index[0, :].long())
+        nodes_sec = torch.index_select(out, 0, edge_label_index[1, :].long())
+        out = torch.sum(nodes_fst * nodes_sec, dim=-1)
+
+        if self.objective == LightGCN.objectives[1]:
+            out = torch.sigmoid(out)
+
+        return out
+
+    def predict_link(self, edge_index: Adj, edge_label_index: Adj = None,
+                     prob: bool = False) -> Tensor:
+        """Predict links between nodes specified in :obj:`edge_label_index`.
+
+        This prediction head works only for the "link_prediction" objective.
+
+        Args:
+            edge_index (Adj): Edge tensor specifying the connectivity of the
+                graph.
+            edge_label_index (Adj, optional): Edge tensor specifying the start
+                and end nodes for which to compute rankings or proabilities.
+                If :obj:`edge_label_index` is set to :obj:`None`, all edges in
+                :obj:`edge_index` will be used (default: :obj:`None`).
+            prob (bool): Whether probabilities should be returned.
+        """
+        assert (self.objective == LightGCN.objectives[1]), 'Prediction ' \
+            'only works for "link_prediction" objective.'
+
+        if edge_label_index is None:
+            edge_label_index = edge_index
+
+        predictions = self.forward(edge_index, edge_label_index)
+        if prob:
+            return predictions
+
+        return torch.round(predictions)
+
+    def recommend(self, edge_index: Adj, source_indices: LongTensor,
+                  target_indices: LongTensor, topK: int = 1) -> Tensor:
+        """Get :obj:`topK` recommendations for nodes in :obj:`source_indices`.
+
+        This prediction head works only for the "ranking" objective.
+
+        Args:
+            edge_index (Adj): Edge tensor specifying the connectivity of the
+                graph.
+            source_indices (LongTensor): Nodes for which recommendations
```

Comment:
Can we make this optional? For example, finding the top-k entries for only a set of source or target nodes (not both).